google/cloud/firestore_bundle/bundle.py

```diff
@@ -0,0 +1,395 @@
+            _read_time = doc.read_time
+        return _read_time
+
+    def _save_named_query(
+        self, name: str, query: BaseQuery, read_time: datetime.datetime,
+    ) -> None:
+        self.named_queries[name] = self._build_named_query(
+            name=name, snapshot=query, read_time=_helpers.build_timestamp(read_time),
+        )
+        self._update_last_read_time(read_time)
+
+    async def _process_async_query(
+        self, name: str, snapshot: AsyncQuery,
+    ) -> datetime.datetime:
+        doc: DocumentSnapshot
+        _read_time = datetime.datetime.min.replace(tzinfo=UTC)
+        async for doc in snapshot.stream():
+            self.add_document(doc, query_name=name)
+            _read_time = doc.read_time
+        return _read_time
+
+    def _build_named_query(
+        self, name: str, snapshot: BaseQuery, read_time: Timestamp,
+    ) -> NamedQuery:
+        return NamedQuery(
+            name=name,
+            bundled_query=BundledQuery(
+                parent=name,
+                structured_query=snapshot._to_protobuf()._pb,
+                limit_type=snapshot.limit_type,
+            ),
+            read_time=read_time,
+        )
+
+    def _update_last_read_time(
+        self, read_time: Union[datetime.datetime, Timestamp]
+    ) -> None:
+        _ts: Timestamp = (
+            read_time
+            if isinstance(read_time, Timestamp)
+            else _datetime_to_pb_timestamp(read_time)
+        )
+
+        # if `_ts` is greater than `self.latest_read_time`
+        if _helpers.compare_timestamps(_ts, self.latest_read_time) == 1:
+            self.latest_read_time = _ts
+
+    def _add_bundle_element(self, bundle_element: BundleElement, *, client: BaseClient, type: str):  # type: ignore noqa
+        """Applies BundleElements to this FirestoreBundle instance as a part of
+        deserializing a FirestoreBundle string.
+        """
+        from google.cloud.firestore_v1.types.document import Document
+
+        if getattr(self, "_doc_metadata_map", None) is None:
+            self._doc_metadata_map = {}
+        if type == "metadata":
+            self._deserialized_metadata = bundle_element.metadata  # type: ignore
+        elif type == "named_query":
+            self.named_queries[bundle_element.named_query.name] = bundle_element.named_query  # type: ignore
+        elif type == "document_metadata":
+            self._doc_metadata_map[
+                bundle_element.document_metadata.name
+            ] = bundle_element.document_metadata
+        elif type == "document":
+            parsed_path = _helpers.parse_reference_value(bundle_element.document.name)
+            snapshot = DocumentSnapshot(
+                data=_helpers.decode_dict(
+                    Document(mapping=bundle_element.document).fields, client
+                ),
+                exists=True,
+                reference=DocumentReference(
+                    parsed_path["collection_name"],
```

Comment:
would it be helpful to make this a concrete type, or a named tuple (currently a dict)

Reply:
I originally shied away from a `NamedTuple` because I normally default to dataclasses (not 3.6-friendly), or attrs (didn't feel worth adding for this). But, a simple class/struct is a good compromise.